### DISCUSSION TITLE
Optional param with linked prefix directory (w.$workspace)

### DISCUSSION THREAD

**Discussion by depsimon** [kiliman/remix-flat-routes#118] [Q&A] [answered]
I'm trying to setup a route that would allow us to handle several cases.

I initially wanted to use `/_app.($workspace).($project)+/index.tsx` which works, but makes it possible for edge cases when workspace or project matches a real page name. Like if the workspace or the project is called `issues` Remix doesn't know what to do on /remix/issues/issues

Basically what I want is something like: `/_app.(w.$workspace).(p.$project)+/issues.tsx` to make sure that remix captures the workspace & project name after the respective `/w/` or `/p/`.

It should resolve the following:

- /issues -> no workspace, no project, show all issues you have access to
- /w/remix/issues -> workspace: remix, no project, show all issues from remix
- /w/remix/p/trellix/issues -> workspace: remix, project: trellix, show all issues from trellix

Ideally it should not accept:
- /p/trellix/issues -> could be handled in a layout loader to make sure the workspace is required, though technically it could be a viable URL.

I have been trying a few options, but I'm not even sure that we can have multiple segments in parenthesis.

**Comment by kiliman** [comment, marked as answer]
Is there a default workspace and project if none is specified?

I'm not a big fan of multiple URLs referring to the same concept. I think optional params really confuse things, especially since RR doesn't do tree-based matching (that is, it doesn't match routes at each URL segment but matches the URL against the entire route config).

See https://github.com/remix-run/remix/issues/9227#issuecomment-2059225888

Anyway, I would first create the explicit route, then create additional routes that redirect to the explicit route.

So start with explicit route `_app.w.$workspace.p.$project+/issues.tsx`

Then, let's say the default workspace/project is `all`.

For `/issues` I would redirect to `/w/all/p/all/issues`

Then create a route `_app.w.$workspace+/issues.tsx` that redirects to `/w/:workspace/p/all/issues`

The nice thing about this is that you can never have a match on `/p/trellix/issues` since there is no route defined.

Anyway, even if you did want to use optional params, RR only allows params to be a single segment (the part between the `/`). So you can't have an optional `(/w/:workspace)`.

**Reply by depsimon** [reply]
Thanks for your thorough answer.

Having a default `all` value for workspace/project actually makes it easier to manage and avoid duplication in routes. I think that's a good solution for our case.
